signingscript/sign.py

```diff
@@ -0,0 +1,434 @@
+    """Sign the internals of a file with the widevine key.
+
+    Extract the zip or tarball and only sign a handful of files (see
+    `_WIDEVINE_BLESSED_FILENAMES` and `_WIDEVINE_UNBLESSED_FILENAMES).
+    The blessed files should be signed with the `widevine_blessed` format.
+    Then recreate the zip or tarball.
+
+    Args:
+        context (SigningContext): the signing context
+        orig_path (str): the source file to sign
+        fmt (str): the format to sign with
+
+    Returns:
+        str: the path to the signed archive
+
+    """
+    file_base, file_extension = os.path.splitext(orig_path)
+    # Extract the archive
+    if file_base.endswith('.tar'):
+        files = await _extract_tarfile(context, orig_path, compression=file_extension)
+    elif file_extension == '.zip':
+        files = await _extract_zipfile(context, orig_path)
+    else:
+        raise SigningScriptError("Unknown widevine file format: {}".format(orig_path))
+    # Sign the appropriate inner files
+    for from_ in files:
+        if _should_sign_widevine(from_):
+            await sign_file(context, from_, fmt)
+            files.append("{}.sig".format(from_))
+    # Recreate the archive
+    if file_extension == '.zip':
+        await _create_zipfile(context, orig_path, files)
+    if file_base.endswith('.tar'):
+        await _create_tarfile(context, orig_path, files, compression=file_extension)
+    return orig_path
+
+
+# _should_sign_windows {{{1
+def _should_sign_windows(filename):
+    """Return True if filename should be signed."""
+    # These should already be signed by Microsoft.
+    _dont_sign = [
+        'D3DCompiler_42.dll', 'd3dx9_42.dll',
+        'D3DCompiler_43.dll', 'd3dx9_43.dll',
+        'msvc*.dll',
+    ]
+    ext = os.path.splitext(filename)[1]
+    b = os.path.basename(filename)
+    if ext in ('.dll', '.exe') and not any(fnmatch.fnmatch(b, p) for p in _dont_sign):
+        return True
+    return False
+
+
+# _should_sign_widevine {{{1
+def _should_sign_widevine(filename):
+    """Return (True, blessed) if filename should be signed."""
+    base_filename = os.path.basename(filename)
+    if base_filename in _WIDEVINE_BLESSED_FILENAMES:
+        return "widevine_blessed"
+    elif base_filename in _WIDEVINE_NONBLESSED_FILENAMES:
+        return "widevine"
+
+
+# log_shas {{{1
+def log_shas(context, paths):
+    """Log the shas of `paths`.
+
+    Args:
+        context (SigningContext): the signing context
+        paths (list): the list of paths to log shas for
+
+    """
+    for path in paths:
+        rel_signed_file = os.path.relpath(path, context.config['work_dir'])
+        log.info("SHA512SUM: %s SIGNED_FILE: %s",
+                 utils.get_hash(path, "sha512"), rel_signed_file)
+        log.info("SHA1SUM: %s SIGNED_FILE: %s",
+                 utils.get_hash(path, "sha1"), rel_signed_file)
+
+
+# zip_align_apk {{{1
+async def zip_align_apk(context, abs_to):
+    """Replace APK with a zip aligned one."""
```

Comment:
Improvement: I now have more context about what zipalign does (unlike back in February). It's basically a binary optimization. We can add more details, like: "Optimize APK for better run-time performance. Necessary if APK is uploaded on Google Play Store. For more information, see: https://developer.android.com/studio/command-line/zipalign.html"